File: examples/typescript/src/simple-tool.ts
#!/usr/bin/env node
/**
 * Simple MCP Server Example - TypeScript
 * ========================================
 *
 * A minimal MCP server that exposes a single "greet" tool.
 * Perfect for understanding the basics of MCP server implementation in TypeScript.
 *
 * Usage:
 *   npm install
 *   npm run build
 *   node dist/simple-tool.js
 *
 * Configuration for Claude Desktop:
 * {
 *   "mcpServers": {
 *     "simple-tool": {
 *       "command": "node",
 *       "args": ["/absolute/path/to/dist/simple-tool.js"]
 *     }
 *   }
 * }
 */

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
  Tool,
} from "@modelcontextprotocol/sdk/types.js";

/**
 * Define the greet tool interface
 */
interface GreetArgs {
  name: string;
  formal?: boolean;
}

/**
 * Create the MCP server instance
 */
const server = new Server(
  {
    name: "simple-tool-server",
    version: "1.0.0",
  },
  {
    capabilities: {
      tools: {},
    },
  }
);

/**
 * Handle the list_tools request
 * This tells the client what tools are available
 */
server.setRequestHandler(ListToolsRequestSchema, async () => {
  return {
    tools: [
      {
        name: "greet",
        description: "Greet someone by name with a friendly message",
        inputSchema: {
          type: "object",
          properties: {
            name: {
              type: "string",
              description: "The name of the person to greet",
            },
            formal: {
              type: "boolean",
              description: "Whether to use formal greeting",
              default: false,
            },
          },
          required: ["name"],
        },
      } as Tool,
    ],
  };
});

/**
 * Handle the call_tool request
 * This executes the requested tool with provided arguments
 */
server.setRequestHandler(CallToolRequestSchema, async (request) => {
  const { name, arguments: args } = request.params;

  if (name === "greet") {
    const { name: personName, formal = false } = args as GreetArgs;

    let greeting: string;
    if (formal) {
      greeting = `Good day, ${personName}. It's a pleasure to meet you.`;
    } else {
      greeting = `Hey ${personName}! Nice to meet you!`;
    }

    return {
      content: [
        {
          type: "text",
          text: greeting,
        },
      ],
    };
  }

  // Handle unknown tool
  throw new Error(`Unknown tool: ${name}`);
});

/**
 * Start the server
 */
async function main() {
  // Use stdio transport for communication
  const transport = new StdioServerTransport();

  // Connect the server to the transport
  await server.connect(transport);

  // Log to stderr (stdout is used for MCP communication)
  console.error("Simple Tool MCP Server running on stdio");
}

// Run the server
main().catch((error) => {
  console.error("Server error:", error);
  process.exit(1);
});

File: examples/python/simple_tool.py
#!/usr/bin/env python3
"""
Simple MCP Server Example - Python
===================================

This is a minimal MCP server that exposes a single "greet" tool.
Perfect for understanding the basics of MCP server implementation.

Usage:
    python simple_tool.py

Configuration for Claude Desktop:
    {
      "mcpServers": {
        "simple-tool": {
          "command": "python",
          "args": ["/absolute/path/to/simple_tool.py"]
        }
      }
    }
"""

import asyncio
import json
from typing import Any
from mcp.server import Server
from mcp.server.stdio import stdio_server
from mcp.types import Tool, TextContent


# Create the MCP server instance
app = Server("simple-tool-server")


@app.list_tools()
async def list_tools() -> list[Tool]:
    """
    Define the tools available to the LLM.

    This function is called by the MCP client to discover what tools
    this server provides.
    """
    return [
        Tool(
            name="greet",
            description="Greet someone by name with a friendly message",
            inputSchema={
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string",
                        "description": "The name of the person to greet"
                    },
                    "formal": {
                        "type": "boolean",
                        "description": "Whether to use formal greeting",
                        "default": False
                    }
                },
                "required": ["name"]
            }
        )
    ]


@app.call_tool()
async def call_tool(name: str, arguments: Any) -> list[TextContent]:
    """
    Handle tool execution requests from the LLM.

    Args:
        name: The name of the tool being called
        arguments: Dictionary of arguments passed to the tool

    Returns:
        List of text content responses
    """
    if name == "greet":
        person_name = arguments.get("name", "friend")
        is_formal = arguments.get("formal", False)

        if is_formal:
            greeting = f"Good day, {person_name}. It's a pleasure to meet you."
        else:
            greeting = f"Hey {person_name}! Nice to meet you!"

        return [TextContent(type="text", text=greeting)]

    # Handle unknown tool
    raise ValueError(f"Unknown tool: {name}")


async def main():
    """
    Main entry point - starts the MCP server with stdio transport.

    The stdio transport communicates via standard input/output,
    which is the most common way to run MCP servers.
    """
    # Use stdio for communication (standard for MCP servers)
    async with stdio_server() as (read_stream, write_stream):
        # Create initialization options
        init_options = app.create_initialization_options()

        # Run the server
        await app.run(
            read_stream,
            write_stream,
            init_options
        )


if __name__ == "__main__":
    # Run the async main function
    asyncio.run(main())
